# cobrascape
Constraint-based modeling for allelic variation (COBRA for fitness landSCAPEs)

1. Infer dominant evolutionary forces (metabolic selection pressures) from large-scale genomics-phenomics data (e.g., GWAS data).
2. Extend allele-phenotype associations to metabolic-phenotype associations.

Installation
~~~~~~~~~~~~
git clone https://github.com/erolkavvas/cobrascape.git
~~~~~~~~~~~~

### Primary scripts 
- Application to large-scale _M. tuberculosis_ antimicrobial resistance (AMR) dataset, as described in (https://rdcu.be/9rHj)
- `01_pairwise_tests.ipynb`
  - Determines pairwise associations between pan-genome alleles and labeled phenotypes.
  - Generates `Supplementary Data File 1`
- `02_ML_ensemble_SVM.ipynb`
  - Performs machine learning (ensemble support vector machine) for selecting groups of alleles that are predictive of the labeled phenotypes.
  - Generates `Supplementary Data File 2`, `Supplementary Data File 3`, and `svm_ensemble_data`
- `03_epistatic_analysis.ipynb`
  - Uses the data generated by `02_ML_ensemble_SVM.ipynb` to select an initial set of gene-gene pairs, and then performs gene-gene logistic regression modeling of these gene-gene pairs to identify statistical significant genetic interactions.
  - Generates `cooccurence_table_excel`, `cooccurence_table_figures`, and `Supplementary Data File 4`
